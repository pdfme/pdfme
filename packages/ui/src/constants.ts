export const DEFAULT_LANG = 'en' as const;

export const DESTROYED_ERR_MSG = '[@pdfme/ui] this instance is already destroyed';

export const SELECTABLE_CLASSNAME = 'selectable';

export const RULER_HEIGHT = 30;

export const PAGE_GAP = 10;

export const LEFT_SIDEBAR_WIDTH = 45;

export const RIGHT_SIDEBAR_WIDTH = 400;

export const BACKGROUND_COLOR = 'rgb(74, 74, 74)';
