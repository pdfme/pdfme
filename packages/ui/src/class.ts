import ReactDOM from 'react-dom/client';
import { DESTROYED_ERR_MSG, DEFAULT_LANG } from './constants.js';
import { debounce } from './helper.js';
import {
  cloneDeep,
  Template,
  Size,
  Lang,
  Font,
  Plugins,
  UIProps,
  UIOptions,
  PreviewProps,
  getDefaultFont,
  checkUIProps,
  checkTemplate,
  checkInputs,
  checkUIOptions,
  checkPreviewProps,
} from '@pdfme/common';
import { builtInPlugins } from '@pdfme/schemas';

export abstract class BaseUIClass {
  protected domContainer!: HTMLElement | null;
  
  protected root: ReactDOM.Root | null = null;

  protected template!: Template;

  protected size!: Size;

  private lang: Lang = DEFAULT_LANG;

  private font: Font = getDefaultFont();

  private pluginsRegistry: Plugins = builtInPlugins;

  private options: UIOptions = {};

  private readonly setSize = debounce(() => {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    this.size = {
      height: this.domContainer.clientHeight || window.innerHeight,
      width: this.domContainer.clientWidth || window.innerWidth,
    };
    this.render();
  }, 100);

  resizeObserver = new ResizeObserver(this.setSize);

  constructor(props: UIProps) {
    checkUIProps(props);

    const { domContainer, template, options = {}, plugins = {} } = props;
    this.domContainer = domContainer;
    this.template = cloneDeep(template);
    this.options = options;
    this.size = {
      height: this.domContainer.clientHeight || window.innerHeight,
      width: this.domContainer.clientWidth || window.innerWidth,
    };
    this.resizeObserver.observe(this.domContainer);

    const { lang, font } = options;
    if (lang) {
      this.lang = lang;
    }
    if (font) {
      this.font = font;
    }

    if (Object.values(plugins).length > 0) {
      this.pluginsRegistry = plugins;
    }
  }

  protected getLang() {
    return this.lang;
  }

  protected getFont() {
    return this.font;
  }

  protected getPluginsRegistry() {
    return this.pluginsRegistry;
  }

  public getOptions() {
    return this.options;
  }

  public getTemplate() {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);

    return this.template;
  }

  public updateTemplate(template: Template) {
    checkTemplate(template);
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);

    this.template = cloneDeep(template);
    this.render();
  }

  public updateOptions(options: UIOptions) {
    checkUIOptions(options);
    const { lang, font } = options || {};

    if (lang) {
      this.lang = lang;
    }
    if (font) {
      this.font = font;
    }
    this.options = Object.assign(this.options, options);
    this.render();
  }

  public destroy() {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    if (this.root) {
      this.root.unmount();
      this.root = null;
    }

    this.resizeObserver.unobserve(this.domContainer);
    this.domContainer = null;
  }

  protected abstract render(): void;
}
export abstract class PreviewUI extends BaseUIClass {
  protected inputs!: { [key: string]: string }[];

  constructor(props: PreviewProps) {
    super(props);
    checkPreviewProps(props);
    this.inputs = convertToStingObjectArray(cloneDeep(props.inputs));
  }

  public getInputs() {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);

    return this.inputs;
  }

  public setInputs(inputs: { [key: string]: string }[]) {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    checkInputs(inputs);

    this.inputs = convertToStingObjectArray(inputs);
    this.render();
  }

  protected abstract render(): void;
}

type DataItem = {
  [key: string]: string | string[][];
};

type StringifiedDataItem = {
  [key: string]: string;
};

function convertToStingObjectArray(data: DataItem[]): StringifiedDataItem[] {
  return data.map((item) => {
    const stringifiedItem: StringifiedDataItem = {};
    Object.keys(item).forEach((key) => {
      const value = item[key];
      if (Array.isArray(value)) {
        stringifiedItem[key] = JSON.stringify(value);
      } else {
        stringifiedItem[key] = value;
      }
    });
    return stringifiedItem;
  });
}
