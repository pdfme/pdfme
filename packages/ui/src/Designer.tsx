import React from 'react';
import ReactDOM from 'react-dom';
import {
  Template,
  DesignerProps,
  checkDesignerProps,
  checkTemplate,
  PDFME_VERSION,
} from '@pdfme/common';
import { BaseUIClass } from './class';
import { DESTROYED_ERR_MSG } from './constants.js';
import DesignerComponent from './components/Designer/index';
import AppContextProvider from './components/AppContextProvider';
import { cloneDeep } from './helper.js';

class Designer extends BaseUIClass {
  private onSaveTemplateCallback?: (template: Template) => void;
  private onChangeTemplateCallback?: (template: Template) => void;
  private pageCursor: number = 0;

  constructor(props: DesignerProps) {
    super(props);
    checkDesignerProps(props);
  }

  public saveTemplate() {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    if (this.onSaveTemplateCallback) {
      this.onSaveTemplateCallback(this.template);
    }
  }

  public updateTemplate(template: Template) {
    checkTemplate(template);
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    this.template = cloneDeep(template);
    if (this.onChangeTemplateCallback) {
      this.onChangeTemplateCallback(template);
    }
    this.render();
  }

  public onSaveTemplate(cb: (template: Template) => void) {
    this.onSaveTemplateCallback = cb;
  }

  public onChangeTemplate(cb: (template: Template) => void) {
    this.onChangeTemplateCallback = cb;
  }
  
  public getPageCursor() {
    return this.pageCursor
  }

  protected render() {
    if (!this.domContainer) throw Error(DESTROYED_ERR_MSG);
    ReactDOM.render(
      <AppContextProvider
        lang={this.getLang()}
        font={this.getFont()}
        plugins={this.getPluginsRegistry()}
        options={this.getOptions()}
      >
        <DesignerComponent
          template={this.template}
          onSaveTemplate={(template) => {
            this.template = template;
            this.template.pdfmeVersion = PDFME_VERSION;
            if (this.onSaveTemplateCallback) {
              this.onSaveTemplateCallback(template);
            }
          }}
          onChangeTemplate={(template) => {
            this.template = template;
            this.template.pdfmeVersion = PDFME_VERSION;
            if (this.onChangeTemplateCallback) {
              this.onChangeTemplateCallback(template);
            }
          }}
          onPageCursorChange={(newPageCursor: number) => {
            this.pageCursor = newPageCursor
          }}
          size={this.size}
        />
      </AppContextProvider>,
      this.domContainer
    );
  }
}

export default Designer;
