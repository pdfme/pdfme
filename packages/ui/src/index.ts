import Designer from './Designer.js';
import Form from './Form.js';
import Viewer from './Viewer.js';

export { Designer, Viewer, Form };
