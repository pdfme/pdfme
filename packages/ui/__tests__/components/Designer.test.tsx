/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, act, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Designer from '../../src/components/Designer/index.js';
import {
  I18nContext,
  FontContext,
  PluginsRegistry,
} from '../../src/contexts';
import { i18n } from '../../src/i18n';
import { SELECTABLE_CLASSNAME } from '../../src/constants';
import { getDefaultFont } from '@pdfme/common';
import { setupUIMock, getSampleTemplate } from '../assets/helper';
import { text, image } from "@pdfme/schemas"

const plugins = { text, image, }

test('Designer snapshot', async () => {
  setupUIMock();
  let container: HTMLElement = document.createElement('a');
  act(() => {
    const { container: c } = render(
      <I18nContext.Provider value={i18n}>
        <FontContext.Provider value={getDefaultFont()}>
          <PluginsRegistry.Provider value={plugins}>
            <Designer
              template={getSampleTemplate()}
              onSaveTemplate={console.log}
              onChangeTemplate={console.log}
              size={{ width: 1200, height: 1200 }}
              onPageCursorChange={console.log}
            />
          </PluginsRegistry.Provider>
        </FontContext.Provider>
      </I18nContext.Provider>
    );
    container = c;
  });

  await waitFor(() => Boolean(container?.getElementsByClassName(SELECTABLE_CLASSNAME)));
  expect(container.firstChild).toMatchSnapshot();
});
