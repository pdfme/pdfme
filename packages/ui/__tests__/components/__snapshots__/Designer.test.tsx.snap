// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Designer snapshot 1`] = `
<div
  style="position: relative; background: rgb(74, 74, 74); width: 1200px; height: 1200px;"
>
  <div
    style="margin: 0px auto; width: 1200px; height: 1200px;"
  >
    <div
      style="position: absolute; bottom: 6%; width: 800px;"
    >
      <div
        style="display: flex; align-items: center; justify-content: space-evenly; position: relative; z-index: 1; left: calc(50% - 75px); width: 150px; height: 40px; box-sizing: border-box; padding: 12px; border-radius: 6px; background-color: rgba(0, 0, 0, 0.45);"
      >
        <div
          style="display: flex; align-items: center;"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="minus"
                class="anticon anticon-minus"
                role="img"
                style="color: rgb(255, 255, 255);"
              >
                <svg
                  aria-hidden="true"
                  data-icon="minus"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </span>
            
          </button>
          <span
            class="ant-typography css-dev-only-do-not-override-gzal6t"
            style="color: rgb(255, 255, 255); font-size: 14px; margin: 8px;"
          >
            <strong>
              100
              %
            </strong>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="plus"
                class="anticon anticon-plus"
                role="img"
                style="color: rgb(255, 255, 255);"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                  />
                  <path
                    d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
                  />
                </svg>
              </span>
            </span>
            
          </button>
        </div>
      </div>
    </div>
    <div
      style="left: 0px; right: 0px; position: absolute; z-index: 1; height: 0px; width: 45px; background: rgb(245, 245, 245); text-align: center; overflow: auto;"
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="draggable"
        role="button"
        tabindex="0"
      >
        <div
          style="visibility: visible;"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-default"
            style="width: 35px; height: 35px; margin-top: 0.25rem; padding: 0.25rem;"
            type="button"
          >
            <div
              style="color: rgba(0, 0, 0, 0.88); display: flex; justify-content: center;"
              title="text"
            >
              <svg
                class="lucide lucide-text-cursor-input"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 4h1a3 3 0 0 1 3 3 3 3 0 0 1 3-3h1"
                />
                <path
                  d="M13 20h-1a3 3 0 0 1-3-3 3 3 0 0 1-3 3H5"
                />
                <path
                  d="M5 16H4a2 2 0 0 1-2-2v-4a2 2 0 0 1 2-2h1"
                />
                <path
                  d="M13 8h7a2 2 0 0 1 2 2v4a2 2 0 0 1-2 2h-7"
                />
                <path
                  d="M9 7v10"
                />
              </svg>
            </div>
            
          </button>
        </div>
      </div>
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="draggable"
        role="button"
        tabindex="0"
      >
        <div
          style="visibility: visible;"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-default"
            style="width: 35px; height: 35px; margin-top: 0.25rem; padding: 0.25rem;"
            type="button"
          >
            <div
              style="color: rgba(0, 0, 0, 0.88); display: flex; justify-content: center;"
              title="image"
            >
              <svg
                class="lucide lucide-image"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <circle
                  cx="9"
                  cy="9"
                  r="2"
                />
                <path
                  d="m21 15-3.086-3.086a2 2 0 0 0-2.828 0L6 21"
                />
              </svg>
            </div>
            
          </button>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; right: 0px; z-index: 1; height: 100%; width: 400px;"
    >
      <div>
        <button
          class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-default ant-btn-icon-only"
          style="position: absolute; display: flex; align-items: center; justify-content: center; top: 1rem; right: 1rem; z-index: 100;"
          type="button"
        >
          <span
            class="ant-btn-icon"
          >
            <span
              aria-label="arrow-right"
              class="anticon anticon-arrow-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="arrow-right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M869 487.8L491.2 159.9c-2.9-2.5-6.6-3.9-10.5-3.9h-88.5c-7.4 0-10.8 9.2-5.2 14l350.2 304H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h585.1L386.9 854c-5.6 4.9-2.2 14 5.2 14h91.5c1.9 0 3.8-.7 5.2-2L869 536.2a32.07 32.07 0 000-48.4z"
                />
              </svg>
            </span>
          </span>
          
        </button>
        <div
          style="width: 400px; height: 100%; display: block; top: 0px; right: 0px; position: absolute; padding: 0.7rem 1rem; overflow-y: auto; font-family: 'Open Sans', sans-serif; box-sizing: border-box; background: rgb(245, 245, 245);"
        >
          <div>
            <div>
              <div
                style="height: 40px; display: flex; align-items: center;"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-gzal6t"
                  style="text-align: center; width: 100%;"
                >
                  <strong>
                    Field List
                  </strong>
                </span>
              </div>
              <div
                class="ant-divider css-dev-only-do-not-override-gzal6t ant-divider-horizontal"
                role="separator"
                style="margin-top: 8px; margin-bottom: 8px;"
              />
              <div
                style="height: 1085px;"
              >
                <div
                  style="height: 100%; overflow-y: auto;"
                >
                  <ul
                    style="margin: 0px; padding: 0px; list-style: none; border-radius: 5px;"
                  >
                    <li
                      style="margin-top: 10px; transform: translate(0px, 0px) scale(1, 1);"
                    >
                      <div
                        style="display: flex; align-items: center; cursor: pointer; border: 1px solid transparent;"
                      >
                        <button
                          class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-default ant-btn-icon-only"
                          style="display: flex; align-items: center; background: none; box-shadow: none; padding-left: 0.25rem;"
                          type="button"
                        >
                          <span
                            class="ant-btn-icon"
                          >
                            <span
                              aria-label="holder"
                              class="anticon anticon-holder"
                              role="img"
                              style="cursor: grab;"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="holder"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M300 276.5a56 56 0 1056-97 56 56 0 00-56 97zm0 284a56 56 0 1056-97 56 56 0 00-56 97zM640 228a56 56 0 10112 0 56 56 0 00-112 0zm0 284a56 56 0 10112 0 56 56 0 00-112 0zM300 844.5a56 56 0 1056-97 56 56 0 00-56 97zM640 796a56 56 0 10112 0 56 56 0 00-112 0z"
                                />
                              </svg>
                            </span>
                          </span>
                          
                        </button>
                        <div
                          style="width: 20px; margin-right: 0.5rem; color: rgba(0, 0, 0, 0.88); display: flex; justify-content: center;"
                          title="text"
                        >
                          <svg
                            class="lucide lucide-text-cursor-input"
                            fill="none"
                            height="20"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M5 4h1a3 3 0 0 1 3 3 3 3 0 0 1 3-3h1"
                            />
                            <path
                              d="M13 20h-1a3 3 0 0 1-3-3 3 3 0 0 1-3 3H5"
                            />
                            <path
                              d="M5 16H4a2 2 0 0 1-2-2v-4a2 2 0 0 1 2-2h1"
                            />
                            <path
                              d="M13 8h7a2 2 0 0 1 2 2v4a2 2 0 0 1-2 2h-7"
                            />
                            <path
                              d="M9 7v10"
                            />
                          </svg>
                        </div>
                        <span
                          class="ant-typography css-dev-only-do-not-override-gzal6t"
                          style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis; width: 100%;"
                          title="Edit"
                        >
                          field1
                        </span>
                      </div>
                    </li>
                    <li
                      style="margin-top: 10px; transform: translate(0px, 0px) scale(1, 1);"
                    >
                      <div
                        style="display: flex; align-items: center; cursor: pointer; border: 1px solid transparent;"
                      >
                        <button
                          class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-default ant-btn-icon-only"
                          style="display: flex; align-items: center; background: none; box-shadow: none; padding-left: 0.25rem;"
                          type="button"
                        >
                          <span
                            class="ant-btn-icon"
                          >
                            <span
                              aria-label="holder"
                              class="anticon anticon-holder"
                              role="img"
                              style="cursor: grab;"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="holder"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M300 276.5a56 56 0 1056-97 56 56 0 00-56 97zm0 284a56 56 0 1056-97 56 56 0 00-56 97zM640 228a56 56 0 10112 0 56 56 0 00-112 0zm0 284a56 56 0 10112 0 56 56 0 00-112 0zM300 844.5a56 56 0 1056-97 56 56 0 00-56 97zM640 796a56 56 0 10112 0 56 56 0 00-112 0z"
                                />
                              </svg>
                            </span>
                          </span>
                          
                        </button>
                        <div
                          style="width: 20px; margin-right: 0.5rem; color: rgba(0, 0, 0, 0.88); display: flex; justify-content: center;"
                          title="image"
                        >
                          <svg
                            class="lucide lucide-image"
                            fill="none"
                            height="20"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <rect
                              height="18"
                              rx="2"
                              ry="2"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <circle
                              cx="9"
                              cy="9"
                              r="2"
                            />
                            <path
                              d="m21 15-3.086-3.086a2 2 0 0 0-2.828 0L6 21"
                            />
                          </svg>
                        </div>
                        <span
                          class="ant-typography css-dev-only-do-not-override-gzal6t"
                          style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis; width: 100%;"
                          title="Edit"
                        >
                          field2
                        </span>
                      </div>
                    </li>
                  </ul>
                </div>
                <div
                  id="DndDescribedBy-1"
                  style="display: none;"
                >
                  
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
                </div>
                <div
                  aria-atomic="true"
                  aria-live="assertive"
                  id="DndLiveRegion-0"
                  role="status"
                  style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
                />
                <div
                  style="padding-top: 0.5rem; display: flex; align-items: center; justify-content: flex-end;"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-gzal6t ant-btn-text ant-btn-sm"
                    type="button"
                  >
                    <u>
                       
                      Bulk update field names
                    </u>
                    
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: relative; overflow: auto; margin-right: 400px; width: 800px; height: 1200px;"
    >
      <div
        class="selecto-selection rCS179bs95 pdfme-selecto"
        style="background-color: rgb(22, 119, 255); opacity: 0.75; border-color: #91caff;"
      />
      <div
        style="transform: scale(1); transform-origin: top left; height: 1px; width: 1px;"
      >
        <div
          style="font-family: 'Roboto'; top: 30px; left: 30px; position: relative; background-image: url(data:image/png;base64,a...); background-size: 793.700787411px 1122.5196850527px; width: 793.700787411px; height: 1122.5196850527px;"
        >
          <div
            class="ruler-container"
            style="position: absolute; top: -30px; left: -30px; height: 30px; width: 30px; background: rgb(51, 51, 51);"
          />
          <div
            class="scena-manager scena-horizontal  rCS1dv2ewo"
            data-styled-id="rCS1dv2ewo"
            style="position: absolute; top: -30px; left: 0px; height: 30px; width: 793.700787411px; background: rgb(51, 51, 51);"
          >
            <div
              class="scena-guide-origin"
            />
            <canvas
              height="0"
              style="width: 100%; height: 100%;"
              width="0"
            />
            <div
              class="scena-guides"
              style="transform: translateY(0px);"
            >
              <div
                class="scena-guide scena-adder"
              />
            </div>
          </div>
          <div
            class="scena-manager scena-vertical  rCS1dv2ewo"
            data-styled-id="rCS1dv2ewo"
            style="position: absolute; top: 0px; left: -30px; height: 1122.5196850527px; width: 30px; background: rgb(51, 51, 51);"
          >
            <div
              class="scena-guide-origin"
            />
            <canvas
              height="0"
              style="width: 100%; height: 100%;"
              width="0"
            />
            <div
              class="scena-guides"
              style="transform: translateX(0px);"
            >
              <div
                class="scena-guide scena-adder"
              />
            </div>
          </div>
          <div
            class="moveable-control-box   pdfme-moveable rCSuehiy8"
            data-able-draggable="true"
            data-able-origin="true"
            data-able-padding="true"
            data-able-resizable="true"
            data-able-rotatable="true"
            data-able-snappable="true"
            data-styled-id="rCSuehiy8"
            style="position: absolute; display: none; transform: translate3d(0px, 0px, 0); --zoom: 1; --zoompx: 1px; --moveable-color: #1677ff;"
          >
            <div
              class="moveable-rotation"
              style="transform: translate(-50%) translate(0px, 0px) rotate(0rad);"
            >
              <div
                class="moveable-line moveable-rotation-line"
                style="transform: scaleX(1); --moveable-color: #1677ff;"
              />
              <div
                class="moveable-control moveable-rotation-control"
                style="transform: translate(0.5px) scale(1);"
              />
            </div>
            <div
              class="moveable-control moveable-direction moveable-n "
              data-direction="n"
              data-rotation="0"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-w "
              data-direction="w"
              data-rotation="90"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-s "
              data-direction="s"
              data-rotation="0"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-e "
              data-direction="e"
              data-rotation="90"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-nw "
              data-direction="nw"
              data-rotation="135"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-ne "
              data-direction="ne"
              data-rotation="45"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-sw "
              data-direction="sw"
              data-rotation="45"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-direction moveable-se "
              data-direction="se"
              data-rotation="135"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-control moveable-origin"
              style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
            />
            <div
              class="moveable-line moveable-direction "
              data-direction=""
              data-line-index="0"
              data-rotation="-1"
              style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px; --moveable-color: #1677ff;"
            />
            <div
              class="moveable-line moveable-direction "
              data-direction=""
              data-line-index="1"
              data-rotation="-1"
              style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px; --moveable-color: #1677ff;"
            />
            <div
              class="moveable-line moveable-direction "
              data-direction=""
              data-line-index="2"
              data-rotation="-1"
              style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px; --moveable-color: #1677ff;"
            />
            <div
              class="moveable-line moveable-direction "
              data-direction=""
              data-line-index="3"
              data-rotation="-1"
              style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px; --moveable-color: #1677ff;"
            />
          </div>
          <div>
            <div
              class="selectable"
              id="1"
              style="position: absolute; cursor: pointer; height: 56.692913386499995px; width: 377.95275591px; top: 75.590551182px; left: 75.590551182px; transform: rotate(0deg); opacity: 1; outline: 1px dashed #1677ff;"
              title="field1"
            >
              <div
                style="height: 100%; width: 100%;"
              >
                <div
                  style="padding: 0px; resize: none; background-color: transparent; display: flex; flex-direction: column; justify-content: flex-start; width: 100%; height: 100%; cursor: default;"
                >
                  <div
                    id="text-1"
                    style="font-family: inherit; color: rgb(0, 0, 0); font-size: 30pt; letter-spacing: 0pt; line-height: 1em; text-align: left; white-space: pre-wrap; word-break: break-word; resize: none; outline: none; margin-bottom: 0px; padding-top: 3.436640625px; background-color: transparent;"
                  >
                    <span
                      style="letter-spacing:inherit;"
                    >
                      b
                    </span>
                    <span
                      style="letter-spacing:0;"
                    >
                      b
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="selectable"
              id="2"
              style="position: absolute; cursor: pointer; height: 151.181102364px; width: 377.95275591px; top: 132.2834645685px; left: 75.590551182px; transform: rotate(0deg); opacity: 1; outline: 1px dashed #1677ff;"
              title="field2"
            >
              <div
                style="height: 100%; width: 100%;"
              >
                <div
                  style="width: 100%; height: 100%; background-image: none; background-size: contain; background-repeat: no-repeat; background-position: center;"
                >
                  <img
                    src="aaaaaaaaaaaa"
                    style="height: 100%; width: 100%; border-radius: 0; object-fit: contain;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      id="DndDescribedBy-0"
      style="display: none;"
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="DndLiveRegion-1"
      role="status"
      style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
    />
  </div>
</div>
`;
